### PR TITLE
TTL for columns and tables

### PR DESCRIPTION
**Pull request by CurtizJ**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

For changelog. Remove if this is non-significant change.

Category (leave one):
- New Feature

Short description (up to few sentences):
Implements TTL for columns and tables.

Detailed description (optional):
TTL is set while creating MergeTree-family table as expression: DateTime column + interval:
`CREATE TABLE t (d DateTime, a Int TTL d + INTERVAL 1 DAY) engine = MergeTree ...`
Then merges are assigned considering TTLs and expired values are removed at needed time while merge executing.

### CODE REVIEW

**Comment by alesapin** [comment]
Abort here https://pastebin.com/TZTyMCcP.

File: dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 1]
```diff
@@ -211,7 +217,10 @@ bool MergeTreeDataMergerMutator::selectPartsToMerge(
         merge_settings.base = 1;
 
     /// NOTE Could allow selection of different merge strategy.
-    merge_selector = std::make_unique<SimpleMergeSelector>(merge_settings);
+    if (has_part_with_expired_ttl)
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp]: What if we get parts with TTL and their values are expires each second? Seems like we will merge only, and touch other parts which may require "normal merge".

File: dbms/src/Storages/MergeTree/MergeTreeData.cpp [hunk 2]
```diff
@@ -156,6 +157,18 @@ MergeTreeData::MergeTreeData(
         min_format_version = MERGE_TREE_DATA_MIN_FORMAT_VERSION_WITH_CUSTOM_PARTITIONING;
     }
 
+    for (const auto & [name, ast] : getColumns().ttl_expressions)
+    {
+        ASTPtr ttl_ast = ast;
+        auto syntax_result = SyntaxAnalyzer(global_context).analyze(ttl_ast, getColumns().getAllPhysical());
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeData.cpp]: Maybe this is not good place for this code. TTL doesn't seem to be specific for MergeTree engines only. Maybe add some virtual method for IStorage?

File: dbms/src/Storages/MergeTree/MergeTreeData.cpp [hunk 3]
```diff
@@ -156,6 +157,18 @@ MergeTreeData::MergeTreeData(
...
+        auto expr = ExpressionAnalyzer(ttl_ast, syntax_result, global_context).getActions(false);
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeData.cpp]: What if exception would be thrown here?
For example correct TTL expression but with absent column. I think we should catch exception, and throw it with error code `BAD_TTL_EXPRESSION`.

File: dbms/src/Storages/MergeTree/MergeTreeData.cpp [hunk 4]
```diff
@@ -1358,6 +1371,21 @@ MergeTreeData::AlterDataPartTransactionPtr MergeTreeData::alterDataPart(
     return transaction;
 }
 
+void MergeTreeData::removeEmptyColumnsFromPart(MergeTreeData::MutableDataPartPtr & data_part)
+{
+    if (!data_part->empty_columns.empty())
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeData.cpp]: Add some logging about such important action.

**Comment by CurtizJ** [comment]
TODO
[] save min-max TTL for specific columns and table

**Comment by alesapin** [comment]
Stress test failure doesn't seem like known error.

File: dbms/src/Storages/MergeTree/TTLMergeSelector.cpp [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+    if (partition_to_merge_index == -1 || partition_to_merge_min_ttl > current_time)
+        return {};
+
+    const auto & best_partition = partitions[partition_to_merge_index];
+    Iterator best_end = best_begin;
+    size_t total_size = 0;
+
+    while (best_begin != best_partition.begin())
+    {
+        if (!best_begin->min_ttl || best_begin->min_ttl > current_time
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/TTLMergeSelector.cpp]: Seems like we will never check this condition for the first part in partition and will always execute TTL on it if previous was suitable.

File: dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp [hunk 6]
```diff
@@ -210,8 +217,17 @@ bool MergeTreeDataMergerMutator::selectPartsToMerge(
...
 
+    time_t merge_with_ttl_timeout = data.global_context.getConfigRef().getInt("merge_with_ttl_timeout", 3600);
+    bool can_merge_with_ttl = (current_time - last_merge_with_ttl > merge_with_ttl_timeout);
+
     /// NOTE Could allow selection of different merge strategy.
-    merge_selector = std::make_unique<SimpleMergeSelector>(merge_settings);
+    if (can_merge_with_ttl && has_part_with_expired_ttl)
+    {
+        merge_selector = std::make_unique<TTLMergeSelector>(current_time);
+        last_merge_with_ttl = current_time;
```
Comment (alesapin) [inline dbms/src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp]: What if merge will take more than 3600 seconds?

**Comment by alesapin** [comment]
@CurtizJ Your own tests failed.

**Pull request merged**